File: tests/python_to_cpp/Shed Skin Examples/4.sudoku4.py
## Solve Every Sudoku Puzzle

## See http://norvig.com/sudoku.html

## Throughout this program we have:
##   r is a row,    e.g. 'A'
##   c is a column, e.g. '3'
##   s is a square, e.g. 'A3'
##   d is a digit,  e.g. '9'
##   u is a unit,   e.g. ['A1','B1','C1','D1','E1','F1','G1','H1','I1']
##   g is a grid,   e.g. 81 non-blank chars, e.g. starting with '.18...7...
##   values is a dict of possible values, e.g. {'A1':'123489', 'A2':'8', ...}

from typing import Dict, Set

def cross(aa, bb):
    return [a+''+b for a in aa for b in bb]

rows = 'ABCDEFGHI'
cols = '123456789'
digits   = '123456789'
squares  = cross(rows, cols)
unitlist = ([cross(rows, str(c)) for c in cols] +
            [cross(str(r), cols) for r in rows] +
            [cross(rs, cs) for rs in ['ABC','DEF','GHI'] for cs in ['123','456','789']])
units = dict([(s, [u for u in unitlist if s in u])
             for s in squares])
peers : Dict[str, Set[str]] = {}
for s in squares:
    se = set() # str
    for u in units[s]:
        for s2 in u:
            if s2 != s:
                se.add(s2)
    peers[s] = se

def search(values) -> Dict[str, str]: # [-TODO: Optional[Dict[str, str]] support (for `return {}` -> `return None`)-]
    "Using depth-first search and propagation, try all possible values."
    if len(values) == 0:
        return {} ## Failed earlier
    if all([len(values[s]) == 1 for s in squares]):
        return values ## Solved!
    ## Chose the unfilled square s with the fewest possibilities
    s = min([(len(values[s]), s) for s in squares if len(values[s]) > 1])[1]
    for d in values[s]:
        global assign
        values_copy = values.copy()
        r = search(assign(values_copy, s, d))
        if len(r) != 0: return r
    return {}

def assign(values : Dict[str, str], s, d) -> Dict[str, str]:
    "Eliminate all the other values (except d) from values[s] and propagate."
    global eliminate
    if all([len(eliminate(values, s, d2)) != 0 for d2 in values[s] if d2 != d]):
        return values
    else:
        return {}

def eliminate(values : Dict[str, str], s, d) -> Dict[str, str]:
    "Eliminate d from values[s]; propagate when values or places <= 2."
    if d not in values[s]:
        return values ## Already eliminated
    values[s] = values[s].replace(d,'')
    if len(values[s]) == 0:
        return {} ## Contradiction: removed last value
    elif len(values[s]) == 1:
        ## If there is only one value (d2) left in square, remove it from peers
        d2 = values[s][0]
        if not all([len(eliminate(values, s2, d2)) != 0 for s2 in peers[s]]):
            return {}
    ## Now check the places where d appears in the units of s
    for u in units[s]:
        dplaces = [s for s in u if d in values[s]]
        if len(dplaces) == 0:
            return {}
        elif len(dplaces) == 1:
            # d can only be in one place in unit; assign it there
            if len(assign(values, dplaces[0], d)) == 0:
                return {}
    return values

def parse_grid(grid) -> Dict[str, str]:
    "Given a string of 81 digits (or .0-), return a dict of {cell:values}"
    grid = [c for c in grid if c in '0.-123456789']
    values = dict([(s, digits) for s in squares]) ## Each square can be any digit
    for s,d in zip(squares, grid):
        if d in digits and len(assign(values, s, d)) == 0:
            return {}
    return values

def solve_file(filename, sep, action):
    "Parse a file into a sequence of 81-char descriptions and solve them."
    results = [action(search(parse_grid(grid)))
               for grid in open(filename).read().strip().split(sep)]
    print("## Got %d out of %d" % (
          sum([(1 if len(r) else 0) for r in results]), len(results))) # [-TODO: `[1 if len(r) else 0 for r in results]` support-]
    return results

def printboard(values):
    "Used for debugging."
    width = 1+max([len(values[s]) for s in squares])
    line = '\n' + '+'.join(['-'*(width*3)]*3)
    for r in rows:
        print(''.join([values[r+c].center(width) + ('|' if c in '36' else '')
                      for c in cols]) + (line if r in 'CF' else ''))
    print()
    return values

if __name__ == '__main__':
    solve_file("testdata/top95.txt", '\n', lambda v: printboard(v)) # [-TODO: `lambda v: printboard(v)` -> `printboard`-]

## References used:
## http://www.scanraid.com/BasicStrategies.htm
## http://www.krazydad.com/blog/2005/09/29/an-index-of-sudoku-strategies/
## http://www2.warwick.ac.uk/fac/sci/moac/currentstudents/peter_cock/python/sudoku/

File: tests/python_to_cpp/Shed Skin Examples/4.sudoku4.cpp
#include "C:\!!BITBUCKET\11l-lang\_11l_to_cpp\11l.hpp"

template <typename T1, typename T2> auto cross(const T1 &aa, const T2 &bb)
{
    return multiloop(aa, bb, [](const auto &a, const auto &b){return a + b;});
}
auto rows = u"ABCDEFGHI"_S;
auto cols = u"123456789"_S;
auto digits = u"123456789"_S;
auto squares = cross(rows, cols);
auto unitlist = (cols.map([](const auto &c){return cross(::rows, String(c));}) + rows.map([](const auto &r){return cross(String(r), ::cols);}) + multiloop(create_array({u"ABC"_S, u"DEF"_S, u"GHI"_S}), create_array({u"123"_S, u"456"_S, u"789"_S}), [](const auto &rs, const auto &cs){return cross(rs, cs);}));
auto units = create_dict(squares.map([](const auto &s){return make_tuple(s, ::unitlist.filter([&s](const auto &u){return in(s, u);}));}));
Dict<String, Set<String>> peers;

struct CodeBlock1
{
    CodeBlock1()
    {
        for (auto &&s : squares) {
            auto se = Set<String>();
            for (auto &&u : units[s])
                for (auto &&s2 : u)
                    if (s2 != s)
                        se.add(s2);
            peers.set(s, se);
        }
    }
} code_block_1;

template <typename T1> Dict<String, String> search(const T1 &values)
{
    u"Using depth-first search and propagation, try all possible values."_S;
    if (values.empty())
        return Dict<String, String>();
    if (all(::squares.map([&values](const auto &s){return values[s].len() == 1;})))
        return values;
    auto s = _get<1>(min(::squares.filter([&values](const auto &s){return values[s].len() > 1;}).map([&values](const auto &s){return make_tuple(values[s].len(), s);})));
    for (auto &&d : values[s]) {
        auto values_copy = values.copy();
        auto r = search(assign(values_copy, s, d));
        if (!r.empty())
            return r;
    }
    return Dict<String, String>();
}

template <typename T2, typename T3> Dict<String, String> assign(Dict<String, String> &values, const T2 &s, const T3 &d)
{
    u"Eliminate all the other values (except d) from values[s] and propagate."_S;
    if (all(values[s].filter([&d](const auto &d2){return d2 != d;}).map([&s, &values](const auto &d2){return !eliminate(values, s, d2).empty();})))
        return values;
    else
        return Dict<String, String>();
}

template <typename T2, typename T3> Dict<String, String> eliminate(Dict<String, String> &values, const T2 &s, const T3 &d)
{
    u"Eliminate d from values[s]; propagate when values or places <= 2."_S;
    if (!in(d, values[s]))
        return values;
    values.set(s, values[s].replace(d, u""_S));
    if (values[s].empty())
        return Dict<String, String>();
    else if (values[s].len() == 1) {
        auto d2 = _get<0>(values[s]);
        if (!all(::peers[s].map([&d2, &values](const auto &s2){return !eliminate(values, s2, d2).empty();})))
            return Dict<String, String>();
    }

    for (auto &&u : ::units[s]) {
        auto dplaces = u.filter([&d, &values](const auto &s){return in(d, values[s]);});
        if (dplaces.empty())
            return Dict<String, String>();
        else if (dplaces.len() == 1) {
            if (assign(values, _get<0>(dplaces), d).empty())
                return Dict<String, String>();
        }
    }
    return values;
}

template <typename T1> Dict<String, String> parse_grid(T1 grid)
{
    u"Given a string of 81 digits (or .0-), return a dict of {cell:values}"_S;
    grid = grid.filter([](const auto &c){return in(c, u"0.-123456789"_S);});
    auto values = create_dict(::squares.map([](const auto &s){return make_tuple(s, ::digits);}));
    for (auto &&[s, d] : zip(::squares, grid))
        if (in(d, ::digits) && assign(values, s, d).empty())
            return Dict<String, String>();
    return values;
}

template <typename T1, typename T2, typename T3> auto solve_file(const T1 &filename, const T2 &sep, const T3 &action)
{
    u"Parse a file into a sequence of 81-char descriptions and solve them."_S;
    auto results = File(filename).read().trim(make_tuple(u" "_S, u"\t"_S, u"\r"_S, u"\n"_S)).split(sep).map([&action](const auto &grid){return action(search(parse_grid(grid)));});
    print(u"## Got #.0 out of #.0"_S.format(sum_map(results, [](const auto &r){return (!r.empty() ? 1 : 0);}), results.len()));
    return results;
}

template <typename T1> auto printboard(const T1 &values)
{
    u"Used for debugging."_S;
    auto width = 1 + max(::squares.map([&values](const auto &s){return values[s].len();}));
    auto line = u"\n"_S + (create_array({u"-"_S * (width * 3)}) * 3).join(u"+"_S);
    for (auto &&r : ::rows)
        print((::cols.map([&r, &values, &width](const auto &c){return values[r + c].center(width) + (in(c, u"36"_S) ? u"|"_S : u""_S);})).join(u""_S) + (in(r, u"CF"_S) ? line : u""_S));
    print();
    return values;
}

int main()
{
    solve_file(u"testdata/top95.txt"_S, u"\n"_S, [](const auto &v){return printboard(v);});
}
